### DISCUSSION TITLE
@native, a new way to write and call plugins

### DISCUSSION THREAD

**Discussion by aparajita** [ionic-team/capacitor#3850] [Ideas]
@dwieeb, @ikeith, @jcesarmobile 

I just released @​​native for Capacitor: https://github.com/aparajita/ws-capacitor-native-decorator. I believe it represents an improvement in the plugin DX, both for plugin authors and for users of plugins.

Looking forward to input. 🙏

**Comment by imhoffd** [comment]
Just summarizing our discussion on Slack:

- **There are some concerns around decorators**, which are not an ECMAScript standard. To make decorators work at runtime, TypeScript polyfills the implementation (which can be tree-shaken with `tslib`). They are an experimental feature of TypeScript and introducing them to the Capacitor API will require extensive research into long-term feasibility. It may make sense to explore alternative implementations for `@native`.
- **When it comes to the unpacking of params and return values**, we'd prefer a more explicit approach vs implicitly checking if number of arguments is one and it's an object.
- **We are trying to avoid building custom tooling for plugin development**, which can incur maintenance cost from the Capacitor core team. We'd like to keep our focus on the experience for Capacitor users. The DX for plugin developers is crucial, but still secondary.

We didn't get to talk about:

- **Hard-coding `@native` to work with promises** may not cover all use cases. See Geolocation's [`watchPosition`](https://capacitorjs.com/docs/apis/geolocation#watchposition) as an example of a non-promise Plugin API method.
- **Web implementations are not loaded on native devices** in Capacitor 3. With `@native`, web implementations _must_ be loaded, which reverts our efforts to limit the dead JS being included on native devices. It certainly makes it nice to interop with native and JS methods in the same class, but we need to be careful about our JS weight. Every byte counts.

**Reply by aparajita** [reply]
Given all these concerns, it seems that `@native` just isn't right for Capacitor's core needs and strategic direction. Better to keep it as a third party plugin rather than increase the support burden on Capacitor.

**Reply by imhoffd** [reply]
@aparajita Would you be willing to add `@native` to [Capacitor community](https://github.com/capacitor-community) (as `@capacitor-community/native-decorator`)? Then we can add a callout in the docs for plugin authors to try it. It would be good to get some real world testing and hear feedback from people before we fully adopt it. What do you think? 😁

**Reply by aparajita** [reply]
Okay, do I need permissions to publish to that org? I don't see myself in the member list.

**Reply by imhoffd** [reply]
I sent you an invite. Connect with me on Slack and we can get you all set up. Thanks!
